src/nagini_translation/analyzer.py

```diff
@@ -460,15 +460,53 @@ def visit_ClassDef(self, node: ast.ClassDef) -> None:
                 if not (isinstance(base, ast.Name) and
                         base.id in MYPY_SUPERCLASSES):
                     actual_bases.append(base)
+
+        if len(actual_bases) > 0:
+            cls.superclass = self.find_or_create_target_class(actual_bases[0])
+
+            # Process ADT defined by a class hierarchy (sum) and a NamedTuple (product)
+            if isinstance(cls.superclass, PythonClass) and cls.superclass.is_adt:
+                cls.is_adt = True
+                # Detect malformed ADTs
+                # ADT classes can have at most two superclasses
+                if len(actual_bases) > 2:
+                    raise InvalidProgramException(cls.node, 'malformed.adt',
+                            'malformed algebraic data type: superclasses can only ' +
+                            'be a class optionally followed by one NamedTuple')
+                # ADT classes should have no body
+                if len(node.body) != 1 or not isinstance(node.body[0], ast.Pass):
+                    raise InvalidProgramException(cls.node, 'malformed.adt',
+                            'malformed algebraic data type: classes cannot have ' +
+                            'body, fields should be defined in NamedTuples instead')
+                elif len(actual_bases) == 2:
+                    # ADT classes can only define fields by inheriting from NamedTuple
+                    if not (isinstance(actual_bases[1], ast.Call)
+                       and actual_bases[1].func.id == 'NamedTuple'):
+                        raise InvalidProgramException(actual_bases[1], 'malformed.adt',
+                            'malformed algebraic data type: only NamedTuple can be ' +
+                            'used to define fields')
+                    # The name of the NamedTuple should match the ADT being defined
```

Comment:
How about "should match the ADT constructor being defined"?
And in the above exception "used to define constructors"?

Reply:
After thinking a bit, it actually makes more sense. Changed.